README.md
# Processing.rb

[ [English](https://github.com/kitao/processing.rb/blob/master/README.md) | [Japanese](https://github.com/kitao/processing.rb/blob/master/README.ja.md) ]

Processing.rb enables you to write various Processing sketches in Ruby easily.

## Features

Compared to similar tools, Processing.rb has the following features:

- Requires no configuration file and available with just adding a module to your Ruby code
- Reloads the sketch file and loaded modules automatically when the related files are updated
- Available to use extension libraries for Processing with just adding a few line code in the same way as Processing
- Provides useful functions for live coding, such as setting the window position and showing it in the topmost

## Screenshots

The following are screenshots of the examples of Processing.rb. Click each image to view the source code.

<a href="https://github.com/kitao/processing.rb/blob/master/examples/01_simple_sketch.rb" target="_blank">
<img
src="https://raw.githubusercontent.com/kitao/processing.rb/master/examples/screenshots/01_simple_sketch.png" width="30%">
</a>
<a href="https://github.com/kitao/processing.rb/blob/master/examples/02_input_handling.rb" target="_blank">
<img
src="https://raw.githubusercontent.com/kitao/processing.rb/master/examples/screenshots/02_input_handling.png" width="30%">
</a>
<a href="https://github.com/kitao/processing.rb/blob/master/examples/03_multi_file.rb" target="_blank">
<img src="https://raw.githubusercontent.com/kitao/processing.rb/master/examples/screenshots/03_multi_file.png" width="30%">
</a>

<a href="https://github.com/kitao/processing.rb/blob/master/examples/04_builtin_library.rb" target="_blank">
<img src="https://raw.githubusercontent.com/kitao/processing.rb/master/examples/screenshots/04_builtin_library.png" width="30%">
</a>
<a href="https://github.com/kitao/processing.rb/blob/master/examples/05_external_library.rb" target="_blank">
<img src="https://raw.githubusercontent.com/kitao/processing.rb/master/examples/screenshots/05_external_library.png" width="30%">
</a>

## How to Install

### Preparations

In order to use Processing.rb, installation of Ruby and Java is required. Each tool can be obtained from the following sites:

- [Ruby](https://www.ruby-lang.org/)
- [Java](https://java.com/)

### Installing Processing.rb

Processing.rb can be installed with the `gem` command of Ruby.

```sh
gem install processing.rb
```

If a permission error occured while installation, please add the `sudo` command at the head of the above command.

In a proxy environment such as in an office, please add a proxy configuration as [-p option](http://guides.rubygems.org/command-reference/#gem-install) at the end of the above command, such as `-p http://proxy.hostname:port`.

After installation, the examples of Processing.rb can be copied to the `~/processingrb_examples` directory with the following command:

```sh
setup_processingrb_examples
```

## How to Use

### Creating a sketch

In Processing.rb, a sketch is created as a derived class from the `Processing::SketchBase` class. And it starts rendering by calling the `Processing.#start` function.

```ruby
require 'processing'

class Sketch < Processing::SketchBase
  def setup
    # implement your own setup code
  end

  def draw
    # implement your own draw code
  end
end

Processing.start(Sketch.new)
```

In the `Processing::SketchBase` class, the constants of Processing in Java such as `HALF_PI` can be used as the same name. Regarding the functions and variables of Processing, their names are [snake-cased](http://en.wikipedia.org/wiki/Snake_case). For example, `no_stroke` in the case of `noStroke` in Java.

Please check the actual code in [the examples](https://github.com/kitao/processing.rb/tree/master/examples).

A sketch file can be run in the same way as a usual Ruby code.

```sh
ruby [sketchfile]
```

Only for the first time, JRuby and Processing will be downloaded to the `~/.processing.rb` directory. The proxy setting to download will be asked at that moment. Please input it or just press `Enter` without any input if not necessary, and wait for a while.

While the sketch file is running, it will be reloaded automatically when the `.rb` files in the same directory are updated.

### Handling input data

The information of the keyboard and mouse can be obtained in the same way as Processing in Java.

But please note that the `keyPressed` and `mousePressed` methods in Java are renamed to `key_pressed?` and `mouse_pressed?` to avoid duplication of the method names.

```ruby
def draw
  if mouse_pressed?
    ellipse(mouse_x, mouse_y, 10, 10)
  end
end

def key_pressed
  Processing.reload if key == 'r'
end
```

This example draws circles when the mouse button is pressed, and restarts the sketch when the `R` key is pressed.

### Using an extension library

An extension library for Processing can be used in Processing.rb in the same way as Processing in Java.

For example, in the case of the sketch uses the video extension library like this:

```java
import processing.video.*;
Movie movie;

void setup() {
  movie = new Movie(this, "sample.mov");
  movie.loop();
}
```

In Processing.rb, it goes as follow:

```ruby
Processing.load_library 'video'
Processing.import_package 'processing.video', 'Video'

class Sketch < Processing::SketchBase
  def setup
    @movie = Video::Movie.new(self, Processing.sketch_path('sample.mov'))
    @movie.loop
  end
```

When a file path for a Java library is specified in Processing.rb, it should be an absolute path. So this example uses the `Processing.#sketch_path` function to convert the relative path from the sketch directory to the absolute path.

In the case of using an extension library not bundled in Processing, please make the `libraries` directory in the same directory as the sketch file and place the library in it.

### Live coding

With passing optional parameters to the `Processing.#start` function, the both an editor screen and a sketch window can get more visible. It helps with editing code in parallel with checking its rendering result.

```ruby
Processing.start(Sketch.new, topmost: true, pos: [300, 300])
```

The options of the `Processing.#start` function are as follows:

|Option|Description|
|----|----|
|topmost:|When its value is `true`, the sketch window always appears in the topmost.|
|pos:|The sketch window appears in the position `[x, y]`.|

## API Reference

### Processing module

The `Processing` module provides the classes and methods for a Processing sketch.

|Constant|Description|
|----|----|
|SKETCH_FILE|The path of the sketch file in the absolute path|
|SKETCH_NAME|The name of the sketch file without the directory|
|SKETCH_DIR|The directory of the sketch file in the absolute path|

|Class|Description|
|----|----|
|SketchBase|The base class of a Processing sketch|

|Method|Description|
|----|----|
|load_library(name)|Loads the specified Processing library|
|load_jars(dir)|Loads all of the `.jar` files in the specified directory|
|import_package(package, module_name)|Imports all of the classes in the specified Java package to the specified module|
|sketch_path(path)|Converts the relative path from the sketch directory to the absolute path|
|start(sketch, topmost: false, pos: nil)|Starts to render the specified sketch instance|
|reload|Reloads the sketch file manually|

## License

Processing.rb is under [MIT license](http://en.wikipedia.org/wiki/MIT_License). Feel free to use it.

Have fun!


CHANGELOG.md
# Change Log

## 1.1.1
- Renamed a constant name
- Modified a error message
- Updated comments in code

## 1.1.0
- Splited the source code into multiple files
- Changed to download Processing automatically
- Changed to start automatically when the module is required
- Renamed complete_path to sketch_path
- Added the utility script to copy the examples
- Adapted to both 32-bit and 64-bit architecture
- Changed to show an error message for downloading files
- Updated the READMEs

## 1.0.0
- Stable release
- Modified the out messages

## 0.9.9
- Fixed a bug with downloading JRuby without proxy
- Modified the out messages

## 0.9.8
- Updated the READMEs
- Modified the .rubocop.yml file
- Fixed a bug with downloading JRuby via proxy

## 0.9.7
- Changed the way to check the downloaded data
- Modified the ouput messages
- Updated the READMEs

## 0.9.6
- Renamed the command name to processing-rb
- Modified the output messages
- Updated the READMEs

## 0.9.5
- Updated the READMEs
- Changed the directories to search for Processing

## 0.9.4
- Enabled to execute from Ruby
- Changed to download JRuby automatically
- Changed the way to find Processing on Windows

## 0.9.3
- Refactored a variable name
- Renamed sketch_path to complete_path

## 0.9.2
- Updated the license text
- Fixed the SketchBase#frame_rate method

## 0.9.1
- Changed the link to download in README
- Renamed examples/assets to examples/data
- Refined the READMEs
- Removed unnecessary comments

## 0.9.0
- Added README in Japanese
- Resized the screenshots of the examples
- Updated README in English
- Refined comments

## 0.8.0
- Updated the example #5
- Refined the example #3
- Added screenshots of the examples

## 0.7.0
- Changed to run with JRuby explicitly
- Modified the example #3
- Updated the examples #4
- Refined comments in the examples

## 0.6.0
- Added an argument to the import_package method
- Included the Handy library for the example #5
- Added the sketch_path method
- Added movie materials for the example #4

## 0.5.0
- Changed the directory structure of the examples
- Renamed load_all_jars to load_jars
- Renamed import_all_classes to import_package
- Updated the example #3
- Fixed the way to set the initial window status

## 0.4.0
- Refactored the example #1
- Implemented the example #2
- Renamed load_jar_files to loar_all_jars
- Added the import_all_classes method

## 0.3.0
- Simplified the example #1
- Changed to include all of the Processing classes in the module
- Enabled to call a class methods as an instance method

## 0.2.0
- Changed the example #1
- Added the .rubocop.yml file
- Renamed start_sketch to start
- Added options to the start method

## 0.1.1
- Refactored the code
- Added a library search path for Linux
- Started implementing the example #1

## 0.1.0
- Changed the way to undefine the Sketch class
- Fixed the frame_rate method
- Refactored the way to restore the environment
- Changed to wrap the whole code in the Processing module
- Added the descriptions to all of the methods

## 0.0.7
- Refactored the way to adapt to the virtual methods
- Refactored the way to execute and watch the sketch
- Wrote the README.md a little

## 0.0.6
- Wrapped the overloaded fields
- Fixed to adapt to the virtual methods in Processing

## 0.0.5
- Added PROCESSING_ROOT check for libraries
- Added a requirement to the gemspec file
- Changed to call jruby when called from ruby
- Fixed to the library load paths in Windows

## 0.0.4
- Renamed a constant variable
- Added a shebang to the main script
- Changed to use JRuby directly instead of Ruby
- Changed the way to find Processing libraries
- Added error messages

## 0.0.3
- Fixed to set __FILE__ in the sketch
- Removed an unnecessary extension in .gitignore

## 0.0.2
- Added a command existence check
- Adapted the load path to the directory of the sketch
- Fixed to catch all of the errors in the sketch

## 0.0.1
- Initial release


lib/sketch_runner/config.rb
# Runs a sketch and reloads it when related files change
module SketchRunner
  VERSION = '1.1.1'

  CONFIG_MTIME = File.stat(__FILE__).mtime

  PACKAGE_ROOT = File.expand_path(File.join(File.dirname(__FILE__), '../..'))

  APPDATA_ROOT = File.expand_path('~/.processing.rb')
  APPDATA_CHECK_FILE = File.join(APPDATA_ROOT, '.complete')

  JRUBY_URL = 'https://s3.amazonaws.com/jruby.org/downloads/9.0.0.0.pre1/jruby-complete-9.0.0.0.pre1.jar'
  JRUBY_FILE = File.join(APPDATA_ROOT, 'jruby/jruby.jar')

  if RUBY_PLATFORM == 'java'
    BIT_SIZE = java.lang.System.getProperty('sun.arch.data.model') == '64' ?
      64 : 32
  else
    require 'open3'
    BIT_SIZE = Open3.capture3('java -version')[1].include?('64-Bit') ? 64 : 32
  end

  if RUBY_PLATFORM == 'java'
    PLATFORM = :JAVA
  elsif /darwin/ =~ RUBY_PLATFORM
    PLATFORM = :MACOSX
  elsif /cygwin|mswin|mingw|bccwin|wince|emx/ =~ RUBY_PLATFORM
    PLATFORM = BIT_SIZE == 32 ? :WIN32 : :WIN64
  else
    PLATFORM = BIT_SIZE == 32 ? :LINUX32 : :LINUX64
  end

  PROCESSING_URL = {
    MACOSX: 'http://download.processing.org/processing-2.2.1-macosx.zip',
    WIN32: 'http://download.processing.org/processing-2.2.1-windows32.zip',
    WIN64: 'http://download.processing.org/processing-2.2.1-windows64.zip',
    LINUX32: 'http://download.processing.org/processing-2.2.1-linux32.tgz',
    LINUX64: 'http://download.processing.org/processing-2.2.1-linux64.tgz'
  }[PLATFORM]

  if PLATFORM == :MACOSX
    PROCESSING_CORE_PATH = 'Processing.app/Contents/Java/core'
    PROCESSING_LIBS_PATH = 'Processing.app/Contents/Java/modes/java/libraries'
  else
    PROCESSING_CORE_PATH = 'processing-2.2.1/core'
    PROCESSING_LIBS_PATH = 'processing-2.2.1/modes/java/libraries'
  end

  PROCESSING_DIR = File.join(APPDATA_ROOT, 'processing')
  PROCESSING_ZIP_DIR = File.join(APPDATA_ROOT, 'processing-zip')
  PROCESSING_ZIP_FILE = File.join(PROCESSING_ZIP_DIR, 'processing.zip')

  EXAMPLES_SRC_DIR = File.join(PACKAGE_ROOT, 'examples')
  EXAMPLES_DEST_DIR = File.expand_path('~/processingrb_examples')

  LOAD_PATH = File.join(PACKAGE_ROOT, 'lib')
  STARTUP_FILE = File.join(PACKAGE_ROOT, 'lib/sketch_runner/runner.rb')

  SKETCH_FILE = File.expand_path(ARGV.length > 0 ? ARGV[0] : '')
  SKETCH_DIR, SKETCH_NAME = File.split(SKETCH_FILE)
  SKETCH_LIBS_DIR = File.join(SKETCH_DIR, 'libraries')

  WATCH_INTERVAL = 0.1
end
